src/api/ApiCall.hh

```diff
@@ -11,21 +11,62 @@ namespace ra {
 namespace api {
 
 enum class ApiResult
-{    
+{
     None = 0,       // unspecified
     Success,        // call was successful
     Error,          // an error occurred
     Failed,         // call was unsuccessful, but didn't error
     Unsupported,    // call is not supported and was not made
+    Incomplete,     // call could not be made at this time
 };
 
 struct ApiRequestBase
 {
+protected:
     template<class TRequest, class TCallback>
     static void CallAsync(const TRequest& request, TCallback&& callback)
     {
         ra::services::ServiceLocator::GetMutable<ra::services::IThreadPool>().RunAsync(
-            [request, callback = std::move(callback)] { callback(request.Call()); });
+            [request, callback = std::move(callback)]{ callback(request.Call()); });
+    }
+
+    template<class TRequest, class TCallback>
+    static void CallAsyncWithRetry(const TRequest& request, TCallback&& callback)
+    {
+        ra::services::ServiceLocator::GetMutable<ra::services::IThreadPool>().RunAsync(
+            [request, callback = std::move(callback)]
+        {
+            DoAsyncWithRetry(std::move(request), std::move(callback), std::chrono::milliseconds(0));
+        });
+    }
+
+private:
+    template<class TRequest, class TCallback>
+    static void DoAsyncWithRetry(const TRequest&& request, TCallback&& callback, std::chrono::milliseconds delay)
+    {
+        auto response = request.Call();
+        if (response.Result != ApiResult::Incomplete)
+        {
+            callback(response);
+            return;
+        }
+
+        if (delay < std::chrono::milliseconds(500))
+        {
+            delay = std::chrono::milliseconds(500);
+        }
+        else
+        {
+            delay += delay;
+            if (delay > std::chrono::minutes(2))
+                delay = std::chrono::minutes(2);
```

Comment:
These look like good opportunities to use `std::chrono_literals`.

Reply:
I prefer not to use suffixes for non-primitives. It's very clear that this is comparing against 2 minutes, whereas "2m" could be potentially be meters or something else.

Reply:
I got you.